// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/explorer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _sidebar = _interopRequireDefault(require("./sidebar"));
var _patternsList = _interopRequireDefault(require("./patterns-list"));
var _blockPatternsTab = require("../block-patterns-tab");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternsExplorer({
  initialCategory,
  rootClientId
}) {
  const [searchValue, setSearchValue] = (0, _element.useState)('');
  const [patternSourceFilter, setPatternSourceFilter] = (0, _element.useState)('all');
  const [selectedCategory, setSelectedCategory] = (0, _element.useState)(initialCategory?.name);
  const patternCategories = (0, _blockPatternsTab.usePatternsCategories)(rootClientId, patternSourceFilter);
  return (0, _element.createElement)("div", {
    className: "block-editor-block-patterns-explorer"
  }, (0, _element.createElement)(_sidebar.default, {
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    onClickCategory: setSelectedCategory,
    searchValue: searchValue,
    setSearchValue: setSearchValue,
    patternSourceFilter: patternSourceFilter,
    setPatternSourceFilter: setPatternSourceFilter
  }), (0, _element.createElement)(_patternsList.default, {
    searchValue: searchValue,
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    patternSourceFilter: patternSourceFilter
  }));
}
function PatternsExplorerModal({
  onModalClose,
  ...restProps
}) {
  return (0, _element.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Patterns'),
    onRequestClose: onModalClose,
    isFullScreen: true
  }, (0, _element.createElement)(PatternsExplorer, {
    ...restProps
  }));
}
var _default = PatternsExplorerModal;
exports.default = _default;
