// @wordpress/block-editor/build/components/block-canvas/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExperimentalBlockCanvas = ExperimentalBlockCanvas;
exports.default = void 0;
var _element = require("@wordpress/element");
var _compose = require("@wordpress/compose");
var _blockList = _interopRequireDefault(require("../block-list"));
var _editorStyles = _interopRequireDefault(require("../editor-styles"));
var _iframe = _interopRequireDefault(require("../iframe"));
var _writingFlow = _interopRequireDefault(require("../writing-flow"));
var _observeTyping = require("../observe-typing");
var _copyHandler = require("../copy-handler");
var _blockSelectionClearer = require("../block-selection-clearer");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ExperimentalBlockCanvas({
  shouldIframe = true,
  height = '300px',
  children = (0, _element.createElement)(_blockList.default, null),
  styles,
  contentRef: contentRefProp,
  iframeProps
}) {
  const resetTypingRef = (0, _observeTyping.useMouseMoveTypingReset)();
  const copyHandler = (0, _copyHandler.useClipboardHandler)();
  const clearerRef = (0, _blockSelectionClearer.useBlockSelectionClearer)();
  const contentRef = (0, _compose.useMergeRefs)([copyHandler, contentRefProp, clearerRef]);
  if (!shouldIframe) {
    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_editorStyles.default, {
      styles: styles,
      scope: ".editor-styles-wrapper"
    }), (0, _element.createElement)(_writingFlow.default, {
      ref: contentRef,
      className: "editor-styles-wrapper",
      tabIndex: -1,
      style: {
        height
      }
    }, children));
  }
  return (0, _element.createElement)(_iframe.default, {
    ...iframeProps,
    ref: resetTypingRef,
    contentRef: contentRef,
    style: {
      width: '100%',
      height,
      ...iframeProps?.style
    },
    name: "editor-canvas"
  }, (0, _element.createElement)(_editorStyles.default, {
    styles: styles
  }), children);
}

/**
 * BlockCanvas component is a component used to display the canvas of the block editor.
 * What we call the canvas is an iframe containing the block list that you can manipulate.
 * The component is also responsible of wiring up all the necessary hooks to enable
 * the keyboard navigation across blocks in the editor and inject content styles into the iframe.
 *
 * @example
 *
 * ```jsx
 * function MyBlockEditor() {
 *   const [ blocks, updateBlocks ] = useState([]);
 *   return (
 *     <BlockEditorProvider
 *       value={ blocks }
 *       onInput={ updateBlocks }
 *       onChange={ persistBlocks }
 *      >
 *        <BlockCanvas height="400px" />
 *      </BlockEditorProvider>
 *    );
 * }
 * ```
 *
 * @param {Object}    props          Component props.
 * @param {string}    props.height   Canvas height, defaults to 300px.
 * @param {Array}     props.styles   Content styles to inject into the iframe.
 * @param {WPElement} props.children Content of the canvas, defaults to the BlockList component.
 * @return {WPElement}               Block Breadcrumb.
 */
function BlockCanvas({
  children,
  height,
  styles
}) {
  return (0, _element.createElement)(ExperimentalBlockCanvas, {
    height: height,
    styles: styles
  }, children);
}
var _default = BlockCanvas;
exports.default = _default;
