// @wordpress/block-editor/build/components/inserter/block-patterns-filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockPatternsSyncFilter = BlockPatternsSyncFilter;
exports.SYNC_TYPES = exports.PATTERN_TYPES = void 0;
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _blockPatternsTab = require("./block-patterns-tab");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const PATTERN_TYPES = {
  all: 'all',
  synced: 'synced',
  unsynced: 'unsynced',
  user: 'user',
  theme: 'theme',
  directory: 'directory'
};
exports.PATTERN_TYPES = PATTERN_TYPES;
const SYNC_TYPES = {
  all: 'all',
  full: 'fully',
  unsynced: 'unsynced'
};
exports.SYNC_TYPES = SYNC_TYPES;
const getShouldDisableSyncFilter = sourceFilter => sourceFilter !== PATTERN_TYPES.all && sourceFilter !== PATTERN_TYPES.user;
const getShouldDisableNonUserSources = category => {
  return category.name === _blockPatternsTab.myPatternsCategory.name;
};
function BlockPatternsSyncFilter({
  setPatternSyncFilter,
  setPatternSourceFilter,
  patternSyncFilter,
  patternSourceFilter,
  scrollContainerRef,
  category
}) {
  // If the category is `myPatterns` then we need to set the source filter to `user`, but
  // we do this by deriving from props rather than calling setPatternSourceFilter otherwise
  // the user may be confused when switching to another category if the haven't explicity set
  // this filter themselves.
  const currentPatternSourceFilter = category.name === _blockPatternsTab.myPatternsCategory.name ? PATTERN_TYPES.user : patternSourceFilter;

  // We need to disable the sync filter option if the source filter is not 'all' or 'user'
  // otherwise applying them will just result in no patterns being shown.
  const shouldDisableSyncFilter = getShouldDisableSyncFilter(currentPatternSourceFilter);

  // We also need to disable the directory and theme source filter options if the category
  // is `myPatterns` otherwise applying them will also just result in no patterns being shown.
  const shouldDisableNonUserSources = getShouldDisableNonUserSources(category);
  const patternSyncMenuOptions = (0, _element.useMemo)(() => [{
    value: SYNC_TYPES.all,
    label: (0, _i18n._x)('All', 'Option that shows all patterns')
  }, {
    value: SYNC_TYPES.full,
    label: (0, _i18n._x)('Synced', 'Option that shows all synchronized patterns'),
    disabled: shouldDisableSyncFilter
  }, {
    value: SYNC_TYPES.unsynced,
    label: (0, _i18n._x)('Not synced', 'Option that shows all patterns that are not synchronized'),
    disabled: shouldDisableSyncFilter
  }], [shouldDisableSyncFilter]);
  const patternSourceMenuOptions = (0, _element.useMemo)(() => [{
    value: PATTERN_TYPES.all,
    label: (0, _i18n.__)('All'),
    disabled: shouldDisableNonUserSources
  }, {
    value: PATTERN_TYPES.directory,
    label: (0, _i18n.__)('Pattern Directory'),
    disabled: shouldDisableNonUserSources
  }, {
    value: PATTERN_TYPES.theme,
    label: (0, _i18n.__)('Theme & Plugins'),
    disabled: shouldDisableNonUserSources
  }, {
    value: PATTERN_TYPES.user,
    label: (0, _i18n.__)('User')
  }], [shouldDisableNonUserSources]);
  function handleSetSourceFilterChange(newSourceFilter) {
    setPatternSourceFilter(newSourceFilter);
    if (getShouldDisableSyncFilter(newSourceFilter)) {
      setPatternSyncFilter(SYNC_TYPES.all);
    }
  }
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.DropdownMenu, {
    popoverProps: {
      placement: 'right-end'
    },
    label: "Filter patterns",
    icon: (0, _element.createElement)(_icons.Icon, {
      icon: (0, _element.createElement)(_components.SVG, {
        width: "24",
        height: "24",
        viewBox: "0 0 24 24",
        fill: "none",
        xmlns: "http://www.w3.org/2000/svg"
      }, (0, _element.createElement)(_components.Path, {
        d: "M10 17.5H14V16H10V17.5ZM6 6V7.5H18V6H6ZM8 12.5H16V11H8V12.5Z",
        fill: "#1E1E1E"
      }))
    })
  }, () => (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.MenuGroup, {
    label: (0, _i18n.__)('Source')
  }, (0, _element.createElement)(_components.MenuItemsChoice, {
    choices: patternSourceMenuOptions,
    onSelect: value => {
      handleSetSourceFilterChange(value);
      scrollContainerRef.current?.scrollTo(0, 0);
    },
    value: currentPatternSourceFilter
  })), (0, _element.createElement)(_components.MenuGroup, {
    label: (0, _i18n.__)('Type')
  }, (0, _element.createElement)(_components.MenuItemsChoice, {
    choices: patternSyncMenuOptions,
    onSelect: value => {
      setPatternSyncFilter(value);
      scrollContainerRef.current?.scrollTo(0, 0);
    },
    value: patternSyncFilter
  })), (0, _element.createElement)("div", {
    className: "block-editor-tool-selector__help"
  }, (0, _element.createInterpolateElement)((0, _i18n.__)('Patterns are available from the <Link>WordPress.org Pattern Directory</Link>, bundled in the active theme, or created by users on this site. Only patterns created on this site can be synced.'), {
    Link: (0, _element.createElement)(_components.ExternalLink, {
      href: (0, _i18n.__)('https://wordpress.org/patterns/')
    })
  })))));
}
